Implement Avatar component

solimant opened the pull request:
Closes #1345 

## ✅ Pull Request Checklist:

- [X] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [X] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [X] Filled out test instructions.
- [X] Updated documentation (if it already exists for this component).
- [X] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:
1. Open the `Avatar` storybook
2. In any of the stories (except the "disabled" one), inspect that `Avatar` can receive focus and show a focus ring

![image](https://user-images.githubusercontent.com/11140840/125377040-2353b580-e341-11eb-8d02-4b3a3a513113.png)

## 🧢 Your Project:
Adobe/RSP

snowystinger commented on packages/@adobe/react-spectrum/package.json:
```diff
@@ -23,6 +23,7 @@
     "@react-aria/ssr": "^3.0.2",
     "@react-aria/visually-hidden": "^3.2.2",
     "@react-spectrum/actiongroup": "^3.2.0",
+    "@react-spectrum/avatar": "3.0.0-alpha.0",
```
ah, no alphas in here

snowystinger commented on packages/@adobe/react-spectrum/src/index.ts:
```diff
@@ -11,6 +11,7 @@
  */
 
 export {ActionGroup} from '@react-spectrum/actiongroup';
+export {Avatar} from '@react-spectrum/avatar';
```
same thing

snowystinger commented on packages/@adobe/spectrum-css-temp/vars/spectrum-large.css:
```diff
@@ -71,6 +71,9 @@
   --spectrum-global-dimension-font-size-1100: 55px;
   --spectrum-global-dimension-font-size-1200: 62px;
   --spectrum-global-dimension-font-size-1300: 70px;
+  --spectrum-alias-avatar-size-100: 26px;
+  --spectrum-alias-avatar-size-400: 36px;
+  --spectrum-alias-avatar-size-600: 46px;
```
these don't have global-dimension values?

snowystinger commented on packages/@adobe/spectrum-css-temp/vars/spectrum-global.css:
```diff
@@ -451,6 +451,12 @@
   --spectrum-alias-heading-xxs-text-size: var(--spectrum-global-dimension-font-size-100);
   --spectrum-alias-heading6-margin-top: var(--spectrum-global-dimension-font-size-75);
   --spectrum-alias-heading-xxs-margin-top: var(--spectrum-global-dimension-font-size-75);
+  --spectrum-alias-avatar-size-50: var(--spectrum-global-dimension-size-200);
```
how'd you decide which ones went in which file?

snowystinger replied:
gotcha, it's the newer stuff, i do not understand how that stuff is organized, resolving

snowystinger commented on packages/@react-spectrum/avatar/chromatic/Avatar.chromatic.tsx:
```diff
@@ -0,0 +1,50 @@
+const SRC_URL_1 = 'https://mir-s3-cdn-cf.behance.net/project_modules/disp/690bc6105945313.5f84bfc9de488.png';
+const SRC_URL_2 = 'https://i.imgur.com/jjiXMsdb.jpg';
+
+storiesOf('Avatar', module)
```
For new stories/chromatic, we try to use CSF, there are some examples in the other chromatic stories
(sorry, plop hasn't been updated yet)

snowystinger commented on packages/@react-spectrum/avatar/docs/Avatar.mdx:
```diff
@@ -0,0 +1,65 @@
+<!-- Copyright 2021 Adobe. All rights reserved.
```
just a heads up but all docs will need to be pulled out into another PR eventually, up to you when you do it

snowystinger commented on packages/@react-spectrum/avatar/package.json:
```diff
@@ -0,0 +1,51 @@
+    "@react-aria/focus": "^3.4.0",
+    "@react-aria/utils": "^3.8.1",
+    "@react-spectrum/utils": "^3.6.1",
+    "@react-stately/collections": "^3.3.2",
```
what're collections used for?

snowystinger commented on packages/@react-spectrum/avatar/src/Avatar.tsx:
```diff
@@ -0,0 +1,60 @@
+          ...styleProps.style,
+          ...(sizeValue && {height: sizeValue, width: sizeValue})
+        }}
+        tabIndex={isDisabled ? undefined : 0} />
```
we'll want an aria package to contain the tabIndex/isDisabled relationship
we should probably contain the alt there as well
not sure if labels will come into play

snowystinger commented on packages/@react-spectrum/avatar/docs/Avatar.mdx:
````diff
@@ -0,0 +1,65 @@
+### Size
+
+```tsx example
+<Avatar alt="Zoe Saldana" size="size-1000" src="https://i.imgur.com/jjiXMsdb.jpg" />
````
I don't think we can use her likeness in our docs, just a guess
we should use the same ones that spectrum.adobe.com has

snowystinger commented on packages/@react-types/avatar/src/index.d.ts:
```diff
@@ -0,0 +1,39 @@
+   /**
+    * The image URL for the avatar.
+    */
+   src: string
```
technically this is optional https://spectrum.adobe.com/page/avatar/#-Use-generic-avatars-when-image-is-undefined

snowystinger replied:
Ah yep, we interpreted it two different ways. Let's get some more opinions.
Down sides of both:
- optional image src means we would need to embed some images by default or code it to a discoverable place that we control so it can't be changed without us knowing
- required img src means potential for people forgetting to implement the generic icon or getting it from the wrong place

snowystinger commented on packages/@react-spectrum/avatar/src/Avatar.tsx:
```diff
@@ -0,0 +1,60 @@
+    : undefined;
+
+  return (
+    <FocusRing focusRingClass={classNames(styles, 'focus-ring')}>
```
if it's focusable, is it also pressable?

snowystinger reviewed (commented):
thanks for getting this going!

solimant reviewed (commented):
@snowystinger - please have another look

snowystinger commented on packages/@react-aria/avatar/test/useAvatar.test.js:
```diff
@@ -0,0 +1,83 @@
+      expect(mockOnPress).toHaveBeenCalled();
+    });
+
+    it('calls and warns when onClick is used', () => {
```
v2 avatar's never officially had onClick, so we shouldn't need to allow for it. We have it on button because it was something new and different compared to previous usages. This is now common for our library so we shouldn't need to allow it anymore

snowystinger commented on packages/@react-spectrum/avatar/src/Avatar.tsx:
```diff
@@ -0,0 +1,64 @@
+export function Avatar(props: SpectrumAvatarProps) {
+  const {
+    isDisabled,
+    size = 'avatar-size-100',
```
I think we can shorten these to their size alone and omit the `avatar-size-` prefix
https://spectrum.adobe.com/page/avatar/#Table-of-options
I could be persuaded to keep the `size-` part since that's how other dimension values are entered in props

snowystinger replied:
actually, do avatar sizes match size?

LFDanLu replied:
+1 to this. Judging from the changes made the global/med/large css files, avatar sizes != spectrum sizes (e.g. `avatar-size-700` = `size-500`). IMO it would easier if we simply redefined the size range provided by Spectrum to be consistent with the standard global sizes (e.g. the valid size value we accept are `size-200` to `size-500`).

snowystinger replied:
hah, unlikely that we're ok with a 1px difference, i think we'll need to bring this up with design and find out how to communicate this best

snowystinger replied:
i did just notice that this is a `size` prop, not `width`, so this may be less of an issue
though maybe we need to remove width/height as props that Avatar accepts?

snowystinger commented on packages/@react-spectrum/avatar/src/Avatar.tsx:
```diff
@@ -0,0 +1,64 @@
+  const {avatarProps} = useAvatar(props, ref);
+  const {alt, ...otherAvatarProps} = avatarProps;
+
+  const {styleProps} = useStyleProps(otherProps);
```
move this above useAvatar, this is usually one of the first hooks we call

snowystinger commented on packages/@react-spectrum/avatar/src/Avatar.tsx:
```diff
@@ -0,0 +1,64 @@
+
+  return (
+    <FocusRing focusRingClass={classNames(styles, 'focus-ring')}>
+      <img
```
what happened to elementType?
https://github.com/adobe/react-spectrum/pull/2104/files#diff-6877c59b7583969ba245039e942bd0a915a3757410b8b2df2e6e56fb3d4360e1R45

snowystinger replied:
can you add some stories that use other elementTypes? can it be svg as well as div? when would it be a div?

LFDanLu replied:
Do we wanna support other element types? The original issue seems to indicate that it should always be a `img`: https://github.com/adobe/react-spectrum/issues/1345

snowystinger replied:
seems fair, i could see someone wanting to support svg pretty quickly

snowystinger commented on packages/@react-spectrum/avatar/stories/Avatar.stories.tsx:
```diff
@@ -0,0 +1,51 @@
+WithAltText.storyName = 'with alt text';
+
+export const CustomSize = AvatarTemplate.bind({});
+CustomSize.args = {...WithAltText.args, size: 'size-1000'};
```
well now i'm confused, is it `avatar-size-1000` or `size-1000`
after looking at the spectrum.adobe.com more, I do see custom sizes, which you are correct should accept dimension values
we'll need to adjust the type for it so that it auto populates type ahead with the avatar sizes, I assume those are the preferential sizes

solimant replied:
Heads up: updated thread link above ☝🏼

snowystinger replied:
yep, resolving this one

LFDanLu commented on packages/@react-aria/avatar/src/useAvatar.ts:
```diff
@@ -0,0 +1,84 @@
+
+  const domProps = filterDOMProps(otherProps);
+
+  const {isPressed, pressProps} = usePress({onPress, onPressStart, onPressEnd, isDisabled, ref});
```
Mentioning this for other opinions:
I know in a previous PR that we made a stance that we only wanted to include isPressed and other states in the aria hook if absolutely necessary. What do we think about having it here? IMO I think we might be able to drop it here

LFDanLu commented on packages/@react-spectrum/avatar/chromatic/Avatar.chromatic.tsx:
```diff
@@ -0,0 +1,46 @@
+Disabled.args = {isDisabled: true, src: SRC_URL_1};
+Disabled.storyName = 'isDisabled';
+
+export const WithAltText = AvatarTemplate.bind({});
+WithAltText.args = {alt: 'Pensive', src: SRC_URL_2};
+WithAltText.storyName = 'with alt text';
```
nit: IMO we could drop this since alt text doesn't modify the look of the component right?

LFDanLu commented on packages/@react-spectrum/avatar/src/Avatar.tsx:
```diff
@@ -0,0 +1,64 @@
+/**
+ * An avatar is a thumbnail representation of an entity, such as a user or an organization.
+ */
+export function Avatar(props: SpectrumAvatarProps) {
```
Should this accept a ref argument?

LFDanLu commented on packages/@react-spectrum/avatar/src/Avatar.tsx:
```diff
@@ -0,0 +1,64 @@
+  const ref = useRef();
+
+  const {avatarProps} = useAvatar(props, ref);
+  const {alt, ...otherAvatarProps} = avatarProps;
```
Does `alt` need to be separated here? Can't you just spread `avatarProps` on img?

snowystinger replied:
that makes me think that avatarProps doesn't have the right return type, it should probably specify that it always returns the alt even if it's just a pass through

LFDanLu commented on packages/@react-spectrum/avatar/stories/Avatar.stories.tsx:
```diff
@@ -0,0 +1,51 @@
+CustomSize.storyName = 'with custom size';
+
+export const Pressable = AvatarTemplate.bind({});
+Pressable.args = {...CustomSize.args, onPress: action('pressed')};
```
Mind adding pressStart and pressEnd here as well?

LFDanLu commented on packages/@react-spectrum/avatar/test/Avatar.test.js:
```diff
@@ -0,0 +1,61 @@
+    expect(screen.getByTestId(/test avatar/i)).toBeInTheDocument();
+  });
+
+  describe('when isDisabled = true', () => {
```
nit: Would be nice to test that press event don't fire/fire depending on if it is disabled or not. Technically handled in the aria tests though heh so not a must do

LFDanLu reviewed (changes requested):
Looks good for the most part, just some comments

solimant reviewed (commented):
@snowystinger, @LFDanLu - thank you for the review. I've addressed your comments/requests. Happy to address any further asks.

LFDanLu commented on packages/@react-spectrum/avatar/package.json:
```diff
@@ -0,0 +1,53 @@
+    "@babel/runtime": "^7.14.6",
+    "@react-aria/avatar": "3.0.0-alpha.0",
+    "@react-aria/interactions": "^3.5.0",
+    "@react-aria/utils": "^3.8.1",
+    "@react-spectrum/utils": "^3.6.1",
```
Think we can remove these now

LFDanLu commented on packages/@react-aria/avatar/package.json:
```diff
@@ -0,0 +1,33 @@
+  },
+  "dependencies": {
+    "@babel/runtime": "^7.6.2",
+    "@react-aria/interactions": "^3.5.0",
```
don't need "@react-aria/interactions" or "@react-types/shared" here either anymore

LFDanLu reviewed (changes requested):
LGTM, just a couple `package.json` cleanup bits

snowystinger reviewed (commented):
I got clarification from design, this is how the 'size' prop should behave

```tsx
<Avatar size="avatar-size-100" /> // allowed
<Avatar size="600px" /> // allowed
<Avatar size="30em" /> // allowed
<Avatar size="100" /> // NOT allowed
<Avatar size="size-100" /> // NOT allowed
<Avatar height="anything" /> // NOT allowed at all
<Avatar width="anything" /> // NOT allowed at all
```

solimant reviewed (commented):
@LFDanLu - I've made the changes, please have another look.

> i did just notice that this is a size prop, not width, so this may be less of an issue
though maybe we need to remove width/height as props that Avatar accepts?

@snowystinger - yes, we're removing `width` and `height` in `SpectrumAvatarProps`.

> I got clarification from design, this is how the 'size' prop should behave

@snowystinger - thank you for checking. I've approached in a way that should satisfy that ask. I'm making an assumption, though, that disallowed values default to the default value; i.e. `avatar-size-100`. Please have another look.

snowystinger commented on packages/@react-aria/avatar/src/useAvatar.ts:
```diff
@@ -0,0 +1,60 @@
+  } = props;
+
+  const avatarProps = {
+    ...(elementType === 'img' && {alt}),
```
I think if it's not labelled, then it's supposed to be aria-role="presentation"
@majornista ?
This is the dom in the default story
<img width="1361" alt="Screen Shot 2021-07-31 at 1 35 48 PM" src="https://user-images.githubusercontent.com/698229/127751906-8478667a-be19-4eb2-83b4-5939bcd2cd40.png">

solimant replied:
As far as screen readers are concerned, having `alt=""` is effectively the same as `role="presentation"`, but the latter is not widely supported:
> Screen readers also allow the use of WAI-ARIA to hide elements by using `role="presentation"`. However, currently, this feature is not as widely supported as using a null `alt` attribute.

Source: https://www.w3.org/WAI/tutorials/images/decorative/#image-used-as-part-of-page-design.

That said, and according to my understanding from the below [excerpt](https://www.w3.org/TR/wai-aria-1.1/#presentation) from WAI-ARIA, we'd want to add a `role="presentation"` if `img` is part of a containing element that provides the role and alternative text of the overall construct:
> Authors _**SHOULD NOT**_ provide meaningful alternative text (for example, use `alt=""` in HTML) when the `presentation` role is applied to an image.
> 
> In the following code sample, the containing `img` and is appropriately labeled by the caption paragraph. In this example the `img` element can be marked as presentation because the role and the text alternatives are provided by the containing element.
> 
> EXAMPLE 10
>  ```html
>  <div role="img" aria-labelledby="caption">
>    <img src="example.png" role="presentation" alt="">
>    <p id="caption">A visible text caption labeling the image.</p>
>  </div>
>  ```

Another thing is, since the `presentation` role causes a given element to be treated as having no role or to be removed from the accessibility tree, do want that for Avatar? It doesn't seem to me that we'd want to omit it from the accessibility tree. I personally see Avatar as being a functional/informative image vs being a decorative image, so if anything, having a prominent role for screen readers is more fitting IMO.

snowystinger replied:
fascinating, thanks for going the extra mile in your research on this one

snowystinger commented on packages/@react-aria/avatar/src/useAvatar.ts:
```diff
@@ -0,0 +1,60 @@
+      'aria-label': alt,
+      role: 'img'
+    }),
+    tabIndex: !isDisabled ? 0 : undefined
```
We need to add focus styles if we're going support it being focusable

solimant replied:
Removed since we agreed that focusable/pressable is not happening now.

snowystinger commented on packages/@react-spectrum/avatar/src/Avatar.tsx:
```diff
@@ -0,0 +1,66 @@
+  const {avatarProps} = useAvatar(props);
+  const {alt, ...otherAvatarProps} = avatarProps; // alt for jsx-a11y/alt-text
+
+  const sizeValue = typeof size !== 'number' && (SIZE_RE.test(size) || !isNaN(size as any))
```
doesn't `typeof size !== 'number'` prevent `!isNaN(size)` from ever being true? or am i confused with double negatives?

solimant replied:
I agree that this part is not super clear. Added a comment.
Basically, we want to cover the case where `size` is a `string` number, such as `"300"` (as opposed to a `number` number; e.g. `300`). In this case, `size` is not of type `number`, but `isNaN` should still return `false`:
```ts
isNaN(300); // false, because 300 is a number
isNaN('300'); // false, because '300' is, essentially, a number
isNaN('300px'); // true, because '300px' is not a number
```
Also, casting `size` as `any` here is a hack since `isNaN` expects `number`, but I want it to handle the `string` number.

LFDanLu replied:
Just wanna make sure: should passing `size="300"` work? Doesn't seem like it does locally

solimant reviewed (commented):
@snowystinger - I've addressed your comments. Please have a look. Thank you 🙏🏼

LFDanLu reviewed (approved):
LGTM, thanks!

snowystinger commented on packages/@react-aria/avatar/src/useAvatar.ts:
```diff
@@ -0,0 +1,58 @@
+ * component. An avatar is a thumbnail representation of an entity, such as a
+ * user or an organization.
+ */
+export function useAvatar(props: AriaAvatarOptions): AvatarAria {
```
we can leave this here for now because the packages are still private and we may decide to move useAvatar into the Avatar component as it may be a very specific to Spectrum component

snowystinger commented on packages/@react-spectrum/avatar/chromatic/Avatar.chromatic.tsx:
```diff
@@ -0,0 +1,42 @@
+Disabled.storyName = 'isDisabled';
+
+export const CustomSize = AvatarTemplate.bind({});
+CustomSize.args = {size: 'size-1000', src: SRC_URL_2};
```
This one is an invalid size now, it'll just do default
```suggestion
CustomSize.args = {size: 100, src: SRC_URL_2};
```

snowystinger reviewed (commented):
One last clean up item and then I'm happy to approve as well

solimant commented:
@snowystinger, @LFDanLu - thank you very much for the review and comments. I've addressed the requested changes, and have also introduced support for more element types via 16f710c. I'm happy to omit those changes and refactor whatever is necessary.

Also, on a similar note, and further to our offline conversations, I understand we want to repurpose `useAvatar` to be something more generic. Perhaps we can call it `useGraphic`, or something of the like..? This can then encapsulate all ARIA-related logic for anything that depends on graphical HTML elements; e.g. `svg`, `img`, etc., and could also be used in the `Image` component. If that makes sense, or if it's at least a good start, I'm happy to pull this into its own PR and refactor what's necessary here to close things off.

LFDanLu commented:
cc @dannify regarding the above. We talked about this offline previously, do we want to go ahead and move the hook's contents into the Avatar component for now? IMO `useGraphic` might be a good name for this hook and we can split it out into another pull to discuss what else the hook may cover.

dannify commented:
I think for the first release I would like to merge the hook into React Spectrum. I think we will need it down the track but we don't have enough of a use case to make a pattern out of it just yet.

solimant commented:
Thank you, @dannify - consolidation is done and ready for review.

snowystinger reviewed (approved):
lgtm! thanks!

LFDanLu reviewed (approved):
LGTM, thanks!
